### ISSUE TITLE
Usage of buffer_time

### ISSUE THREAD

**Issue by wangxiangyu** [Yelp/elastalert#1313]
hi
 I have questions about the usage of buffer_time.
 I read from the doc
```
buffer_time: ElastAlert will continuously query against a window from the present to buffer_time ago. This way, logs can be back filled up to a certain extent and ElastAlert will still process the events. 
```
what does ` back filled up` mean?  update log or index log?
If  'index log',   it means  log doesn’t get indexed immediately, so I can use `query_delay`
if 'update log', in my opinion,  log doesn't need updated once indexed.
Any ideas?

**Comment by Qmando** [comment]
back fill in these case refers to documents uploaded where the @timestamp is behind the current time.

For example:
if ``{"@timestamp": "11:50"}`` was uploaded at 12:00, you'd need a buffer_time of at least 10 minutes to catch this document.

This will never occur if Logstash adds the @timestamp when it processes the document, but can occur if Logstash is parsing logs from another source that contain their own timestamps.

The issue with having a HUGE buffer_time then is that each query is more network traffic and elastalert uses more memory to keep track of events.

Query_delay achieves a similar goal, you can have smaller query windows but they are shifted backwards in time.

**Comment by ryanpersaud** [comment]
> This will never occur if Logstash adds the @timestamp when it processes the document, but can occur if Logstash is parsing logs from another source that contain their own timestamps.

Even with Logstash adding @timestamp, isn't it possible for there to be a delay between when Logstash creates @timestamp and when  the document is actually indexed that could result in a document being missed by an elastalert query?  See: https://github.com/elastic/elasticsearch/issues/15644 for an exhaustive discussion.

**Comment by Qmando** [comment]
Yes, you are right "never" is not really accurate. And in fact, that's why ``query_delay`` is basically necessary if you're using count or terms query, because those don't have overlapping queries.

I suppose what I meant in this context is that you'll never have to worry about changing ``buffer_time`` from the default value if your documents are being timestamped by logstash.

**Comment by Qmando** [comment]
Blacklist is a bad example since that's a *rule*type, not a query type. In fact, you can't use it with anything other than the default.

The other query types are ``use_count_query``, ``use_terms_query``. With these ones, there is no overlap, so in your example, D2 would be missed. You pretty much have to use ``query_delay``. We can't have overlapping queries here because there's no way remove duplicate values, which we normally do by storing all the _ids of processed documents.

Metric aggregations are similar, but they do actually can have overlapping queries, since aggregation values (think average) can be computed over overlapping windows, and it's not "duplicating" anything.
